### DISCUSSION TITLE
'인스턴스로서의 가치'가 어떤 것을 의미할까요?

### DISCUSSION THREAD

**Discussion by moonheekim0118** [woowacourse-study/2022-code-review-study#44] [Youtube] [answered]
> 전체 코드를 둘러보니, Display의 구현클래스들도 실상은 인스턴스로서의 가치가 없이 오직 초기화 용도로만 클래스를 활용하고 있더군요. 이런 경우에까지 클래스를 사용하는 것이 맞는지 고민해보시면 좋겠어요. by 로이님.. - [바로가기](https://github.com/woowacourse/javascript-youtube-classroom/pull/95#discussion_r825242959)

- 저와 페어가 추상 클래스를 작성했는데요, 여러 함수에서 공통으로 사용되고, 공통적으로 실행되어야 하는 함수들을 해당 추상 클래스에서 모아주었습니다. 그래서 반복을 없애고자 했는데요, 로이님이 위와 같은 리뷰를 남겨주셨습니다.
- 여기서 '인스턴스로서의 가치'가 뭘까요? 
- 반복을 줄이고, 여러 클래스 끼리 같은 로직을 공유하고자 작성하는 추상클래스도 '인스턴스로서의 가치'가 필요할까요?
- 단순히 반복을 줄이고자 작성하는 추상클래스는 옳지 못할까요?
- 그렇다면 반복을 줄이는 추상 클래스를 작성하는 것 대신, 여러 클래스가 뚱뚱해지는게 더 좋은 코드일까요 ?

여러분들의 생각은 어떠신가요? @.@...

**Comment by byhhh2** [comment, marked as answer]
현재는 코드가 많이 바뀌어서 [아래 commit](https://github.com/woowacourse/javascript-youtube-classroom/commit/d036e09e6bdf849d05c749e57351f5f5c53b0dbb)만 참고했는데요! 

```diff
export default class Display {
  container;
+  drawList = [];

  constructor() {
    if (new.target.name === Display.name) {
    this.setContainer();
    this.defaultElement();
+    this.setRenderList();
    this.bindEvents();
    this.subscribeStores();
  }

  setContainer() {}

+  setRenderList() {}

  defaultElement() {}

  bindEvents() {}

  subscribeStores() {}

+ render(state) {
+   this.drawList.forEach(drawEvent => {
+   drawEvent(state);
+   });
+ }

+ addDrawList(drawEvent) {
+    this.drawList.push(drawEvent);
+  }

  addEvent(eventType, selector, handler, option = {}) {
    const children = [...this.container.querySelectorAll(selector)];
    const isTarget = target => children.includes(target) || target.closest(selector);
```

추가되기 전 Display코드는 공통되는 함수`이름만` 존재하는 느낌이더라구요. + 그 아이들을 실행해주는 것을 반복하지 않기 위해 생성자에서 실행
저는 로이님의 피드백을 요런 느낌으로 이해했어요

> Display의 구현클래스들도 실상은 인스턴스로서의 가치가 없이(=Display) 오직 초기화 용도로만 클래스(= Display)를 활용하고 있더군요. 

= 공통되는 함수 이름만 추상클래스로 빼는 것은 Display의 의미가 적어지는 느낌인 것 같습니다. 

다른 언어에서 추상클래스를 사용하는 이유는 자식클래스가 추상클래스의 메서드를 반드시 구현하도록 강제하잖아요? 하지만 js는 강제하지 않습니다. (진짜 추상 클래스가 없기 때문에) 이럴경우 더더욱 함수 이름만 존재하는 추상클래스는 필요없는 것이 아닐까요? (+ 생성자에서 실행 반복을 피하기 위해) 그렇기 때문에 display의 가치가 떨어지는 것 같아요.

이건 저의 정말 개인적인 의견이라 정답인지 전혀 모르겠네요. 댓글로 지적해주세요 !

**Comment by wonsss** [comment]
ㅠㅠㅠ

**Comment by juunzzi** [comment]
### 그래서 추상화해보니깐 어땠어?

**장점**

1. **확장**이 쉬울 것 같았어요. 

		만약 새로운 Video List UI가 생긴다면 `VideoListComponent`를 확장 후 추가적인 로직만을 추가해주면 되니깐요

2. 다른 방식에서 **사용되는 로직(메소드,프로퍼티)를 가지고 있을 필요가 없어요.**
		
        각자 클래스에서만 가지고 있으면 되니깐요
        
3. 하나의 **클래스 파일이 커지는 것을 방지**할 수 있어요
		
        하나의 클래스 파일이 무거워지는 것은 가독성을 떨어뜨리는 일인것 같아요. 또 유지보수에도 많은 `side effect`가 따라오지 않을까 싶네요.

4. 각자의 로직만을 수정해야한다면 더 편하겠죠? 
		
        공통 로직은 추상화된 클래스에 있을테니 이 부분을 수정하면되고, 확장된 UI에서 변경사항이 발생해야한다면 그 부분만 수정하면 되니 유지보수에는 유리할 것 같아요 ! 
        
**단점**

1. **구조가 너무 복잡해**
	
    	내가 만약 이 코드를 처음보는 입장이라면 이해하기 힘들었을지도 모르겠다. 
        
        (이를 쉽게 납득하기 위한 구조도가 있지 않은 이상...
        
        또 확장되는 UI가 더 많다면, 이 모든 UI들의 공통 로직을 뽑아내기에 이해하기 더 힘들었겠지..)

> 어떤 리뷰어의 리뷰(정답 아님) : 모든 로직을 전부 확장성 있게 짜면 오래걸리고 구조가 복잡해집니다. 딱히 바뀔 일이 없는 로직을 억지로 확장성 있게 짤 필요는 없습니다. 가능하다면 단순한 게 베스트죠. 추상화가 들어가는 순간 복잡도가 올라갑니다. 항상 트레이드 오프를 고민해야합니다.



2. 섣부르게 한 추상화라 그런지 **정리가 안된 느낌이 너무 쌔**
	
    	처음부터 추상화를 해두고 개발한 것이 아닌, 하나의 파일을 공통로직과 각자의 로직을 분리하여
        
        공통 로직만을 추상화한 경우여서 추상화의 기준이 코드에서 안보이는 느낌?

### 그렇다면 언제 추상화를 고려할 것 같아?

추상화가 무조건 정답이 아닌 것은 이번 미션을 통해 잘 깨달은 것 같아요. 하지만 장점도 가지고 있으니 언제쯤 사용하게 될지 정리해볼게요

- 확장이 자주될 UI이다
	
    
   	 확장이 되지 않을 UI라면 추상화를 고려하지는 않을 것 같아요. 하지만 UI의 재사용이 잦고, 재사용에 로직 확장이 자주 발생한다면 **개발에 빠른 시간을 가져가고**, **각자의 로직을 분리하기 위해** 추상화를 고려할 것 같아요.

- 같은 `속성, 역할`을 가졌다라면 추상화 해볼지도?? 개념적으로 대체될 수 없다면 추상화를 위한 추상화이다!!

  > 두 객체가 동일한 "역할"을 수행하는지 충분히 고민해보고 추상화한다.
동일한 역할은 대체 가능을 의미합니다. 개념적으로 대체될 수 없는 두 객체를 추상화 하는 건 추상화를 위한 추상화가 될 수 있습니다.

**Reply by wonsss** [reply]
> 같은 속성, 역할을 가졌다라면 추상화 해볼지도?? 개념적으로 대체될 수 없다면 추상화를 위한 추상화이다!!

동의
